node/src/components/deploy_acceptor.rs

```diff
@@ -101,120 +184,586 @@ impl DeployAcceptor {
+                } else {
+                    effect_builder
+                        .immediately()
+                        .event(move |_| Event::VerifyPaymentLogic {
+                            event_metadata,
+                            prestate_hash,
+                            verification_start_timestamp,
+                        })
+                }
+            }
+        }
+    }
 
-                return effects;
+    fn verify_payment_logic<REv: ReactorEventT>(
+        &self,
+        effect_builder: EffectBuilder<REv>,
+        event_metadata: EventMetadata,
+        prestate_hash: Digest,
+        verification_start_timestamp: Timestamp,
+    ) -> Effects<Event> {
+        let executable_deploy_item = event_metadata.deploy().payment();
+        let mut maybe_failure: Option<DeployParameterFailure> = None;
+        if let ExecutableDeployItem::Transfer { .. } = executable_deploy_item {
+            maybe_failure = Some(DeployParameterFailure::InvalidPaymentVariant);
+        }
+
+        if let ExecutableDeployItem::ModuleBytes { module_bytes, args } = executable_deploy_item {
+            if !module_bytes.is_empty() {
+                if let Some(value) = args.get(ARG_AMOUNT) {
+                    let _ = value.clone().into_t::<U512>().map_err(|_| {
+                        maybe_failure = Some(DeployParameterFailure::FailedToParsePaymentAmount)
+                    });
+                } else {
+                    maybe_failure = Some(DeployParameterFailure::MissingPaymentAmount)
+                }
             }
+        }
 
-            Some(false) => {
-                self.metrics.observe_rejected(verification_start_timestamp);
-                info! {
-                    "Received deploy from account {} that does not have minimum balance required", account_key
-                };
-                // The client has submitted a deploy from an account that does not have minimum
-                // balance required. Return an error message to the RPC component via the responder.
-                if let Some(responder) = maybe_responder {
-                    effects.extend(responder.respond(Err(Error::InsufficientBalance)).ignore());
+        match maybe_failure {
+            Some(failure) => {
+                effect_builder
+                    .immediately()
+                    .event(move |_| Event::InvalidDeployResult {
+                        event_metadata,
+                        error: Error::InvalidDeployParameters {
+                            prestate_hash,
+                            failure,
+                        },
+                        verification_start_timestamp,
+                    })
+            }
+            None => {
+                let executable_deploy_item_identifier = executable_deploy_item.identifier();
+                let account_hash = event_metadata.deploy().header().account().to_account_hash();
+                match executable_deploy_item_identifier {
```

Comment:
Isn't `executable_deploy_item_identifier` coming from the payment code? Few lines above we've identified that `Transfer` is invalid variant for the payment while `Module` with empty bytes has to have a payment argument. Why do we pattern match on it here again but not check their content and proceed to verification of the session logic?